dandi/download.py

```diff
@@ -607,24 +602,54 @@ def _download_file(
         elif existing is DownloadExisting.REFRESH:
             if op.lexists(annex_path):
                 raise RuntimeError("Not refreshing path in git annex repository")
+
+            # If we have no expected mtime, warn and check file digests if present
             if mtime is None:
-                lgr.warning(
-                    f"{path!r} - no mtime or ctime in the record, redownloading"
-                )
+                if digests is not None and check_digests(path, digests):
+                    lgr.debug("already exists - matching digest")
+                    yield _skip_file("already exists")
+                    return
+                else:
+                    lgr.warning(
+                        f"{path!r} - no mtime or ctime in the record and digests don't match, "
+                        f"redownloading"
+                    )
+
+            # Otherwise, first check against size and mtime because that's fast.
+            # if they don't match, always redownload.
+            # If they do match, only redownload if we have digests and they don't match
             else:
                 stat = os.stat(op.realpath(path))
                 same = []
                 if is_same_time(stat.st_mtime, mtime):
                     same.append("mtime")
                 if size is not None and stat.st_size == size:
                     same.append("size")
-                # TODO: use digests if available? or if e.g. size is identical
-                # but mtime is different
                 if same == ["mtime", "size"]:
                     # TODO: add recording and handling of .nwb object_id
-                    yield _skip_file("same time and size")
-                    return
-                lgr.debug(f"{path!r} - same attributes: {same}.  Redownloading")
+                    # if we have digests, check those before deciding not to redownload
+                    if digests is not None and check_digests(path, digests):
+                        lgr.debug("already exists - same time, size, and digest")
+                        yield _skip_file("already exists - same time, size, and digest")
+                        return
+
+                    # if we don't have digests but size and mtime match, don't redownload
+                    elif digests is None or len(digests) == 0:
```

Comment:
You already set `digests` to `None` if zero-length above.

Reply:
oop sry thought i changed all of those